versioning/BRM/extentmap.cpp

```diff
@@ -269,19 +270,306 @@ FreeListImpl::FreeListImpl(unsigned key, off_t size, bool readOnly) :
 {
 }
 
+/*static*/
+std::mutex ExtentMapIndexImpl::fInstanceMutex_;
+
+/*static*/
+ExtentMapIndexImpl* ExtentMapIndexImpl::fInstance_ = nullptr;
+
+/*static*/
+ExtentMapIndexImpl* ExtentMapIndexImpl::makeExtentMapIndexImpl(unsigned key, off_t size, bool readOnly)
+{
+    std::lock_guard<std::mutex> lock(fInstanceMutex_);
+
+    if (fInstance_)
+    {
+        if (size != fInstance_->getShmemSize())
+        {
+            fInstance_->fBRMManagedShmMemImpl_.remap();
+        }
+
+        return fInstance_;
+    }
+
+    fInstance_ = new ExtentMapIndexImpl(key, size, readOnly);
+    fInstance_->createExtentMapIndexIfNeeded();
+
+    return fInstance_;
+}
+
+ExtentMapIndexImpl::ExtentMapIndexImpl(unsigned key, off_t size, bool readOnly) :
+    fBRMManagedShmMemImpl_(key, size, readOnly)
+{
+}
+
+void ExtentMapIndexImpl::createExtentMapIndexIfNeeded()
+{
+    // pair<T*, size>
+    // Hide fShmSegment
+    auto managedShmemSearchPair = fBRMManagedShmMemImpl_.getManagedSegment()->find<ExtentMapIndex>(EmIndexObjectName);
```

Comment:
Nit, `find_or_construct` here?

Reply:
Looks more natural indeed.
I have two concerns though: what if ExtentMapIndex instance is 0 and this is a major change. I don't want to have reasons to go though another round of full QA.